File: Python/mlarocca/datastructures/heap/dway_heap.py
from typing import Any, List, Optional, Tuple


class DWayHeap(object):
    """Implementation of a d-ary heap.
    The branching factor for the heap can be passed as an argument.
    It's 2 by default, which is also the minimum possible value.
    The branching factor is the maximum number of children that each internal node can have.
    For regular heaps, a node an have at most 2 children, so the branching factor is 2.
    The higher the branching factor, the shortest the height of the heap. However, when an element is
    pushed towards the leaves of the heap, at each step all children of current node must be examined,
    so a larger branching factor implies a higher number of nodes to be checked for each step of this
    operation.
    On the other hand, inserting elements only examines at most h element, where h is the height of the heap,
    so this operation is only made faster with larger branching factors.
    In general values between 3 and 5 are a good compromise and produce good performance."""

    def __init__(self, elements: List[Any] = [], priorities: List[float] = [], branching_factor: int = 2) -> None:
        """Constructor

        Args:
            elements: The elements for initializing the heap.
            priorities: The priorities of the elements above. Must have the same length as `elements`.
            branching_factor: The (max) number of children for each node in the heap. Must be at least 2.
        """
        if len(elements) != len(priorities):
            raise ValueError(f'The length of the elements list ({len(elements)})'
                             f' must match the length of the priorities list ({len(priorities)}).')
        if branching_factor < 2:
            raise ValueError(f'Branching factor ({branching_factor}) must be greater than 1.')
        self._pairs: List[Tuple[float, Any]] = []
        self.D = branching_factor

        if len(elements) > 0:
            self._heapify(elements, priorities)

    def __sizeof__(self) -> int:
        """Size of the heap.

        Returns: The number of elements in the heap.
        """
        return len(self)

    def __len__(self) -> int:
        """Size of the heap.

        Returns: The number of elements in the heap.
        """
        return len(self._pairs)

    def _validate(self) -> bool:
        """Checks that the three invariants for heaps are abided by.
        1.	Every node has at most `D` children. (Guaranteed by construction)
        2.	The heap tree is complete and left-adjusted.(Also guaranteed by construction)
        3.	Every node holds the highest priority in the subtree rooted at that node.

        Returns: True if all the heap invariants are met.
        """
        current_index = 0
        first_leaf = self.first_leaf_index()
        while current_index < first_leaf:
            current_priority: float = self._pairs[current_index][0]
            first_child = self._first_child_index(current_index)
            last_child_guard = min(first_child + self.D, len(self))
            for child_index in range(first_child, last_child_guard):
                if current_priority < self._pairs[child_index][0]:
                    return False
            current_index += 1
        return True

    def _push_down(self, index: int) -> None:
        """Pushes down the root of a sub-heap towards its leaves to reinstate heap invariants.
        If any of the children of the element has a higher priority, then it swaps current
        element with its highest-priority child C, and recursively checks the sub-heap previously rooted
        at that C.

        Args:
            index: The index of the root of the sub-heap.
        """

        # INVARIANT: 0 <= index < n
        assert (0 <= index < len(self._pairs))
        input_pair = self._pairs[index]
        input_priority = input_pair[0]
        current_index = index
        first_leaf = self.first_leaf_index()
        while current_index < first_leaf:
            child_index = self._highest_priority_child_index(current_index)
            assert (child_index is not None)
            if self._pairs[child_index][0] > input_priority:
                self._pairs[current_index] = self._pairs[child_index]
                current_index = child_index
            else:
                break

        self._pairs[current_index] = input_pair

    def _bubble_up(self, index: int) -> None:
        """Bubbles up towards the root an element, to reinstate heap's invariants.
        If the parent P of an element has lower priority, then swaps current element and its parent,
        and then recursively check the position previously held by the P.

        Args:
            index: The index of the element to bubble up.
        """
        # INVARIANT: 0 <= index < n
        assert (0 <= index < len(self._pairs))
        input_pair = self._pairs[index]
        input_priority = input_pair[0]
        while index > 0:
            parent_index = self._parent_index(index)
            parent = self._pairs[parent_index]

            if input_priority > parent[0]:
                self._pairs[index] = parent
                index = parent_index
            else:
                break

        self._pairs[index] = input_pair

    def _first_child_index(self, index) -> int:
        """Computes the index of the first child of a heap node.

        Args:
            index: The index of current node, for which we need to find children's indices.

        Returns: The index of the left-most child for current heap node.
        """
        return index * self.D + 1

    def _parent_index(self, index) -> int:
        """Computes the index of the parent of a heap node.

        Args:
            index: The index of current node, for which we need to find its parent's indices.

        Returns: The index of the parent of current heap node.

        """
        return (index - 1) // self.D

    def _highest_priority_child_index(self, index) -> Optional[int]:
        """Finds, among the children of a heap node, the one child with highest priority.
        In case multiple children have the same priority, the left-most is returned.

        Args:
            index: The index of the heap node whose children are searched.

        Returns: The index of the child of current heap node with highest priority, or None if
                 current node has no child.
        """
        first_index = self._first_child_index(index)
        size = len(self)
        last_index = min(first_index + self.D, size)

        if first_index >= size:
            return None

        highest_priority = -float('inf')
        index = first_index
        for i in range(first_index, last_index):
            if self._pairs[i][0] > highest_priority:
                highest_priority = self._pairs[i][0]
                index = i

        return index

    def first_leaf_index(self):
        return (len(self) - 2) // self.D + 1

    def _heapify(self, elements: List[Any], priorities: List[float]) -> None:
        """Initializes the heap with a list of elements and priorities.

        Args:
            elements: The list of elements to add to the heap.
            priorities: The priorities for those elements (in the same order they are presented).
        """
        assert (len(elements) == len(priorities))
        self._pairs = list(zip(priorities, elements))
        last_inner_node_index = self.first_leaf_index() - 1
        for index in range(last_inner_node_index, -1, -1):
            self._push_down(index)

    def is_empty(self) -> bool:
        """Checks if the heap is empty.

        Returns: True if the heap is empty.

        """
        return len(self) == 0

    def top(self) -> Any:
        """Removes and returns the highest-priority element in the heap.
        If the heap is empty, raises a `RuntimeError`.

        Returns: The element with highest priority in the heap.
        """
        if self.is_empty():
            raise RuntimeError('Method top called on an empty heap.')
        if len(self) == 1:
            element = self._pairs.pop()[1]
        else:
            element = self._pairs[0][1]
            self._pairs[0] = self._pairs.pop()
            self._push_down(0)

        return element

    def peek(self) -> Any:
        """Removes, WITHOUT removing it, the highest-priority element in the heap.
        If the heap is empty, raises a `RuntimeError`.

        Returns: The element with highest priority in the heap.
        """
        if self.is_empty():
            raise RuntimeError('Method peek called on an empty heap.')
        return self._pairs[0][1]

    def insert(self, element: Any, priority: float) -> None:
        """Add a new element/priority pair to the heap

        Args:
            element: The new element to add.
            priority: The priority associated with the new element
        """
        self._pairs.append((priority, element))
        self._bubble_up(len(self._pairs) - 1)

File: JavaScript/src/dway_heap/dway_heap.js
const _branchFactor = new WeakMap();
const _elements = new WeakMap();
const _positions = new WeakMap();
const _compare = new WeakMap();
const _bubbleUp = new WeakMap();
const _pushDown = new WeakMap();
const _setElementToPosition = new WeakMap();
const _removeElementFromPosition = new WeakMap();
const _getElementPositions = new WeakMap();
//DEBUG   const _check = new WeakMap();

function defaultCompare(x, y) {
  if (x < y) {
    return -1;
  } else if (x > y) {
    return 1;
  } else {
    return 0;
  }
}

/**
 * @name validateElem
 * @for DWayHeap
 * @private
 * @description
 *
 * Validate an element.
 *
 * @param {*} value
 * @returns {boolean} True iff it is possible to store this value in the heap
 */
function validateElem(value) {
  return typeof value !== 'undefined' && typeof value !== 'function' && value !== null;
}

/**
 * @name setElementToPosition
 * @for DWayHeap
 * @private
 * @description
 *
 * Set an element position in the heap, updating its location, if necessary.
 * (We must keep track of the location of each element to speed up the updatePriority method,
 *   which is heavily used in Prim and Dijkstra algorithms).
 *
 * @param {!Array} elements The array of elements in the heap.
 * @param {!Map} positions A map for each (unique) elem in the to the positions in which it is stored.
 * @param {!*} elem  The element to be stored in the heap.
 * @param {!Number} index   The index of the new position of elem in the heap array.
 * @param {?Number} oldIndex  Optional parameter: The old index of elem inside the array.
 *
 * @return {undefined}
 * @throws {TypeError(ERROR_MSG_DWAYHEAP_SET_ELEMENT)} If the index parameter is not valid.
 */
function setElementToPosition(elements, positions, elem, index, oldIndex) {
  //TODO remove and check in the callers

  if (!positions.has(elem)) {
    positions.set(elem, [index]);
  } else {
    oldIndex = parseInt(oldIndex, 10);
    if (!Number.isNaN(oldIndex)) {
      //We have to remove exactly one occurrence of the previous position
      let i = positions.get(elem).indexOf(oldIndex);
      if (i >= 0) {
        //The position was actually present, so we have to remove the element at that index
        positions.get(elem).splice(i, 1);
      }
    }
    positions.get(elem).push(index);
  }

  elements[index] = elem;
}

/**
 * @name removeElementFromPosition
 * @for DWayHeap
 * @private
 * @description
 *
 * Remove an element at a certain position in the heap, updating its location, if necessary.
 * (We must keep track of the location of each element to speed up the updatePriority method,
 * which is heavily used in Prim and Dijkstra algorithms).
 *
 * @param {!Array} elements The array of elements in the heap.
 * @param {!Map} positions A map for each (unique) elem in the to the positions in which it is stored.
 * @param {!Number} index   The index of the new position of elem in the heap array.
 * @param {?Boolean} splice   Optional parameter: flag to state if the element must also be esponged from the array.
 *
 * @return {*} The element stored at index.
 * @throws {TypeError(ERROR_MSG_DWAYHEAP_REMOVE_ELEMENT)} If the index parameter is not valid.
 */
function removeElementFromPosition(elements, positions, index, splice) {
  const elem = elements[index];

  //we have to remove exactly one occurrence of the previous position
  const i = positions.get(elem).indexOf(index);
  if (i >= 0) {
    //The position was actually present, so we have to remove the element at that index
    positions.get(elem).splice(i, 1);
  }
  if (splice) {
    //Actually splice the element from the array
    elements.splice(index, 1);  //assert === elem
  }
  return elem;
}

/**
 * @name getElementPositions
 * @for DWayHeap
 * @private
 *
 * @description
 *
 * Get the position of the passed element in the heap.
 * (We must keep track of the location of each element to speed up the updatePriority method,
 * which is heavily used in Prim and Dijkstra algorithms).
 *
 * @param {!Map} positions A map for each (unique) elem in the to the positions in which it is stored.
 * @param {!*} elem  The element whose position is to be retrieved.
 *
 * @return {Number}  The position of the element in the heap elements array. If the same element appears
 *                   in more than one positions, the first one in the list is returned (not necessarily
 *                   the smallest one).
 *                   If the element is not contained in the heap, returns -1.
 */
function getElementPositions(positions, elem) {
  if (positions.has(elem) && positions.get(elem).length > 0) {
    return positions.get(elem);
  } else {
    return null;
  }
}

/**
 * @name bubbleUp
 * @for DWayHeap
 * @private
 * @description
 *
 * Reinstate the properties of the heap by bubbling an element up to the root
 *
 * @param {!Array} elements The array of elements in the heap.
 * @param {!Map} positions A map for each (unique) elem in the to the positions in which it is stored.
 * @param {!Number} branchFactor The branching factor of the heap.
 * @param {!Function} compare A comparator for heap's elements.
 * @param {!Number} index The index of the element to move towards the root of the heap.
 *
 * @return {Number} The new position for the element.
 */
function bubbleUp(elements, positions, branchFactor, compare, index) {
  const elem = elements[index];
  let i = index;
  while (i > 0) {
    let parentIndex = Math.floor((i - 1) / branchFactor);
    if (compare(elem, elements[parentIndex]) < 0) {
      //elem must keep bubbling up: sets elements[i] = elements[parentIndex];
      setElementToPosition(elements, positions, elements[parentIndex], i, parentIndex);
      i = parentIndex;
    } else {
      //we found the right place for elem: position i
      break;
    }
  }
  if (i !== index) {
    //Sets elements[i] = elem;
    setElementToPosition(elements, positions, elem, i, index);
  }
  return i;
}

/**
 * @name pushDown
 * @for DWayHeap
 * @private
 * @description
 *
 * Reinstate the properties of the heap by pushing down an element towards the leaves
 *
 * @param {!Array} elements The array of elements in the heap.
 * @param {!Map} positions A map for each (unique) elem in the to the positions in which it is stored.
 * @param {!Number} branchFactor The branching factor of the heap.
 * @param {!Function} compare A comparator for heap's elements.
 * @param {Number} index The index of the element to push down the heap.
 *
 * @return {Number} The new position for the element.
 */
function pushDown(elements, positions, branchFactor, compare, index) {
  const elem = elements[index];
  const n = elements.length;
  let parentIndex = index;
  let smallestChildIndex = index * branchFactor + 1;
  //Iteratively go down the subtree
  while (smallestChildIndex < n) {
    let smallestChild = elements[smallestChildIndex];
    let m = Math.min(n, smallestChildIndex + branchFactor);
    //Look for the smallest child
    for (let i = smallestChildIndex + 1; i < m; i++) {
      if (compare(elements[i], smallestChild) < 0) {
        smallestChildIndex = i;
        smallestChild = elements[i];
      }
    }
    //Check if the smallest of the children is smaller than elem
    if (compare(smallestChild, elem) < 0) {
      //we need to push the parent down
      setElementToPosition(elements, positions, smallestChild, parentIndex, smallestChildIndex);
      //Iteratively go down the subtree
      parentIndex = smallestChildIndex;
      smallestChildIndex = parentIndex * branchFactor + 1;
    } else {
      //We don't need to go down this subtree
      break;
    }
  }
  if (index !== parentIndex) {
    setElementToPosition(elements, positions, elem, parentIndex, index);
  }

  return parentIndex;
}

/**
 * @name heapify
 * @for DWayHeap
 * @private
 * @description
 *
 * Initialize a heap with a set of elements.
 * INVARIANT: elements contains been checked before being passed to heapify,
 *            and MUST be a valid non-empty array.
 *
 * @param {Array} intialElements The initial set of elements to be added to the heap.
 * @return {undefined}
 */
function heapify(elements, positions, branchFactor, compare, intialElements) {

  const n = intialElements.length;

  for (let i = 0; i < n; i++) {
    setElementToPosition(elements, positions, intialElements[i], i);
  }
  let lastInnerNode = Math.floor((n - 1) / branchFactor);
  for (let i = lastInnerNode; i >= 0; i--) {
    pushDown(elements, positions, branchFactor, compare, i);
  }

  //TODO: ~remove~
  check(elements, branchFactor, compare);
}

function check(elements, branchFactor, compare) {
  const n = elements.length;

  for (let parentIndex = 0; parentIndex < n; parentIndex++) {
    let parent = elements[parentIndex];
    let childIndex = parentIndex * branchFactor + 1;
    let m = Math.min(n, childIndex + branchFactor);
    for (; childIndex < m; childIndex++) {
      if (compare(elements[childIndex], parent) < 0) {
        throw new Error(ERROR_MSG_DWAYHEAP_CHECK());
      }
    }
  }
  return true;
}

const ERROR_MSG_DWAYHEAP_CONSTRUCTOR_FST_PARAM = (val) => `Illegal argument for DWayHeap constructor: branchFactor ${val}`;
const ERROR_MSG_DWAYHEAP_CONSTRUCTOR_SND_PARAM = (val) => `Illegal argument for DWayHeap constructor: elements ${val}`;
const ERROR_MSG_DWAYHEAP_CONSTRUCTOR_TRD_PARAM = (val) => `Illegal argument for DWayHeap constructor: compare ${val}`;
const ERROR_MSG_DWAYHEAP_PUSH = (val) => `Illegal argument for push: ${val}`;
const ERROR_MSG_DWAYHEAP_EMPTY_HEAP = () => `Invalid Status: Empty Heap`;
const ERROR_MSG_DWAYHEAP_CHECK = () => `Heap Properties Violated`;
const ERROR_MSG_DWAYHEAP_UPDATE_PRIORITY = (val) => `Out of range argument: element ${val} not stored in the heap`;
const ERROR_MSG_DWAYHEAP_UPDATE_PRIORITY_API = (val) => `Illegal argument for updatePriority: ${val}`;
const ERROR_MSG_DWAYHEAP_ELEMENT_POSITION = (val) => `Error: can't find position for elem:  ${val}`;

/**
 * @class DWayHeap
 *
 * A d-way heap (aka d-ary heap or d-heap) is a priority queue data structure, a generalization of the binary
 * heap in which the nodes have d children instead of 2. Thus, a binary heap is a 2-heap.
 * D-way heaps are pretty useful in practice in the implementation of Dijkstra and Prim algorithms
 * for graphs, among many other things.
 * While Fibonacci's heaps would be theoretically faster, no simple and fast implementation of such data structures is known.
 * In practice, a 4-way heap is the best solution for the priority queues in these algorithms.
 *
 * This implementation uses a Map to store keys positions inside the heap.
 * This has 3 important consequences:
 * 1. You can store both literals and objects in the heap.
 * 2. If you store an object, to retrieve it you need to pass to .get() the same object. Different objects, even if
 *    containing the same fields, are not considered equal. This is due to a limitation of Maps
 *    (see http://www.2ality.com/2015/01/es6-maps-sets.html).
 * 3. References to the objects stored are held until they are removed from the heap. If we used a WeakMap instead,
 *    references would have been held weakly, allowing garbage collector to dispose those objects if not used elsewhere.
 *    So, for example:
 *
 *      const map = new Map();
 *      let x = [], y = [];
 *      map.set(x, 1);
 *      console.log(map.has(x)); // Will print false
 *
 *    WeekMaps, however, doesn't allow primitive values, so heap keys would have had to be objects.
 */
class DWayHeap {
  constructor(branchFactor = 2, elements = [], compare = defaultCompare) {
    //Force casting to Number and takes the integer part
    const bF = parseInt(branchFactor, 10);
    if (Number.isNaN(bF) || bF < 2) {
      throw new TypeError(ERROR_MSG_DWAYHEAP_CONSTRUCTOR_FST_PARAM(branchFactor));
    }

    if (elements === undefined || elements === null || !Array.isArray(elements)) {
      throw new TypeError(ERROR_MSG_DWAYHEAP_CONSTRUCTOR_SND_PARAM(elements));
    }

    if (typeof compare !== 'undefined' && (typeof compare !== 'function' || compare.length !== 2)) {
      throw new TypeError(ERROR_MSG_DWAYHEAP_CONSTRUCTOR_TRD_PARAM(compare));
    }

    _branchFactor.set(this, bF);
    _compare.set(this, compare);
    _positions.set(this, new Map());
    _elements.set(this, []);

    _bubbleUp.set(this, bubbleUp.bind(this, _elements.get(this), _positions.get(this), bF, _compare.get(this)));
    _pushDown.set(this, pushDown.bind(this, _elements.get(this), _positions.get(this), bF, _compare.get(this)));
    _setElementToPosition.set(this, setElementToPosition.bind(this, _elements.get(this), _positions.get(this)));
    _removeElementFromPosition.set(this, removeElementFromPosition.bind(this, _elements.get(this), _positions.get(this)));
    _getElementPositions.set(this, getElementPositions.bind(this, _positions.get(this)));
    //DEBUG    _check.set(this, check.bind(this, _elements.get(this), bF, _compare.get(this)));

    heapify(_elements.get(this), _positions.get(this), bF, _compare.get(this), elements);
    //DEBUG    _check.get(this)();
  }

  /**
   * @name size
   * @for DWayHeap
   * @getter
   * @description
   *
   * Return the number of elements stored in the heap.
   *
   * @return {Number} The number of elements in the heap.
   */
  get size() {
    return _elements.get(this).length;
  }

  /**
   * @name isEmpty
   * @for DWayHeap
   * @description
   *
   * Check if the heap is empty.
   *
   * @return {Boolean} True <=> the heap is empty, false otherwise.
   */
  isEmpty() {
    return this.size === 0;
  }

  /**
   * @name branchFactor
   * @for DWayHeap
   * @getter
   * @description
   * Return the branch factor of current heap.
   *
   * @returns {number} The branch factor for this heap.
   */
  get branchFactor() {
    return _branchFactor.get(this);
  }


  /**
   * @name contains
   * @for DWayHeap
   *
   * Returns true if the element is stored in the heap, false otherwise
   *
   * @param {!*} elem  The element to look for.
   *
   * @return {Boolean}   true <=> The element is stored in the heap.
   */
  contains(elem) {
    let ps = _getElementPositions.get(this)(elem);
    return ps !== null && ps.length > 0;
  }


  /**
   * @name push
   * @for DWayHeap
   *
   * Add an element to the heap, taking care of reinstating heap's properties.
   *
   * @param {*} elem   The element to add to the heap. Must be a comparable element.
   *
   * @return {Object} The heap itself, to comply with the chaining pattern.
   * @throws {TypeError(ERROR_MSG_DWAYHEAP_PUSH)}  Throws an error if the first parameter is undefined or null.
   */
  push(elem) {
    const n = this.size;

    if (!validateElem(elem)) {
      //The only three types that can't be accepted
      throw new TypeError(ERROR_MSG_DWAYHEAP_PUSH(elem));
    }

    //Add one element to the array
    _setElementToPosition.get(this)(elem, n);
    //Then bubble it up
    _bubbleUp.get(this)(n);

    //DEBUG    _check.get(this)();

    return this;
  }

  /**
   * @name peek
   * @for DWayHeap
   *
   * If the heap is not empty, return a reference to the first element in the heap.
   *
   * @return {*} A deep copy of the top element in the heap
   * @throws {Error(ERROR_MSG_DWAYHEAP_EMPTY_HEAP)}  Throws an error if the heap is empty.
   */
  peek() {
    if (this.isEmpty()) {
      throw new Error(ERROR_MSG_DWAYHEAP_EMPTY_HEAP());
    }
    //deep copy
    return JSON.parse(JSON.stringify(_elements.get(this)[0]));
  }

  /**
   * @name top
   * @for DWayHeap
   *
   * If the heap is not empty, return the first element in the heap, after removing it from the data structure;
   * The heap properties are then reinstated.
   *
   * @return {*} A reference to the top element in the heap
   * @throws {Error(ERROR_MSG_DWAYHEAP_EMPTY_HEAP)}  Throws an error if the heap is empty.
   */
  top() {
    const n = this.size;

    switch (n) {
      case 0:
        throw new Error(ERROR_MSG_DWAYHEAP_EMPTY_HEAP());
      case 1:
        return _removeElementFromPosition.get(this)(0, true);
      default:
        //We MUST not change the position of subsequent array elements
        let topElem = _removeElementFromPosition.get(this)(0, false);
        //The last element, instead, needs to be removed
        let elem = _removeElementFromPosition.get(this)(n - 1, true);
        _setElementToPosition.get(this)(elem, 0, n - 1);
        _pushDown.get(this)(0);

        //DEBUG      _check.get(this)();

        return topElem;
    }
  }

  /**
   * @name updatePriority
   * @for DWayHeap
   *
   * Updates all the element stored in the heap matching oldValue, possibly changing its priority;
   * then it takes care of reinstating heap's properties.
   * If the new priority is greater (i.e. newValue < oldValue) the new element will be pushed toward the root,
   * if it's smaller (i.e. newValue > oldValue) it will be pushed towards the leaves.
   * **WARNING** Be advised that all the occurrences of the old value will be replaced with the instance of the new value passed.
   *             In case `newValue` is an object, the same reference will replace all the occurrences of oldValue.
   *
   * @param {*} oldValue   The element to be updated. MUST be in the heap.
   * @param {*} newValue   The new value for the element.
   *
   * @return {Object} The heap itself, to comply with the chaining pattern.
   * @throws {RangeError(ERROR_MSG_DWAYHEAP_UPDATE_PRIORITY)}  Throws an error if the first parameter is undefined or null.
   * @throws {RangeError(ERROR_MSG_DWAYHEAP_ELEMENT_POSITION)}  Throws an error if the first parameter is undefined or null.
   */
  updatePriority(oldValue, newValue) {
    const compareResult = _compare.get(this)(newValue, oldValue);
    const n = this.size;
    if (!this.contains(oldValue)) {
      //TODO
      throw new RangeError(ERROR_MSG_DWAYHEAP_ELEMENT_POSITION(oldValue));
    }

    let indices = _getElementPositions.get(this)(oldValue);

    if (!validateElem(newValue)) {
      //The only three types that can't be accepted
      throw new TypeError(ERROR_MSG_DWAYHEAP_UPDATE_PRIORITY_API(newValue));
    }
    //First update all the elements
    for (let i of indices) {
      if (i < 0 || i > n) {
        throw new RangeError(ERROR_MSG_DWAYHEAP_UPDATE_PRIORITY(newValue));
      }
      //else
      _setElementToPosition.get(this)(newValue, i);
    }
    //Then we need to restore priority constraints. The order in which we process the elements depends on the operation we need to perform
    if (compareResult < 0) {
      //Priority decreases so the element will have to move up to the root. We have to start from smaller indices.
      indices = indices.sort((x, y) => x - y);
      for (let i of indices) {
        _bubbleUp.get(this)(i);
      }
    } else if (compareResult > 0) {
      //Priority decreases so the element will have to move down to leaves. We have to start from larger indices.
      indices = indices.sort((x, y) => y - x);
      for (let i of indices) {
        _pushDown.get(this)(i);
      }
    }
    //else: nothing to do, priority is the same as before

    //DEBUG    _check.get(this)();

    return this;
  }

  /**
   * @name sorted
   * @for DWayHeap
   *
   * Return a sorted array with all the elements in the heap.
   * WARNING: all the elements will be removed from the heap!
   *
   * @return  {Array} The sorted array with all elements (possibility empty).
   */
  sorted() {
    let res = [];
    while (this.size > 0) {
      res.push(this.top());
    }

    return res;
  }

  /**
   * @name iterator
   * @for DWayHeap
   *
   * Return an iterator that will go through all the elements in the heap, in sorted order
   * **WARNING**: elements will be removed from the heap!
   *
   * @return  {Array} The sorted array with all elements (possibility empty).
   */
  *[Symbol.iterator]() {
    while (this.size > 0) {
      yield this.top();
    }
  }
}

export default DWayHeap;